### DISCUSSION TITLE
Static code analysis for better code quality

### DISCUSSION THREAD

**Discussion by lejcik** [OpenSalamander/salamander#40] [General]
MSVC compiler can do [static code analysis](https://learn.microsoft.com/en-us/cpp/code-quality/code-analysis-for-c-cpp-overview?view=msvc-170), go to menu 'Debug' and 'Run Code Analysis' on a solution (or rather on a project, as it takes long time to finish). The project codebase is quite old, and static code analysis can find additional errors or vulnerabilities in the code that may be dangerous (buffer overruns, leaked resources, etc.). I think it should be worth to try it out.

Github also comes with something similar - [code scanning with CodeQL](https://docs.github.com/en/code-security/code-scanning/introduction-to-code-scanning/about-code-scanning-with-codeql) which is free for open-source projects and supports C/C++.

Also [SonarQube](https://www.sonarsource.com/open-source-editions/sonarqube-community-edition/) may be a good option.

Does anybody have skills with setting up such tools for open-source projects? I'd try to run the analysis on the code manually and will fix the found issues with pull requests, but we have to synchronize with translators, let we don't have unwanted git conflicts...

**Comment by janrysavy** [comment]
Another useful tool is [Clang-Tidy](https://clang.llvm.org/extra/clang-tidy/). Both Clang-Tidy and MSVC static code analysis are based on the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines), created by [Bjarne Stroustrup](http://www.stroustrup.com/) and [Herb Sutter](http://herbsutter.com/). At Fine, we are using both Clang-Tidy and MSVC static code analysis for our C++ codebase.

**Comment by janrysavy** [comment]
Note that Clang-Tidy is capable of automatically fixing and refactoring code. Works really well.

**Comment by lejcik** [comment]
yes, I know `clang-tidy`, we use it in our company for linux-based projects with _cmake_, I wasn't sure if I can use it with [MSBuild projects](https://learn.microsoft.com/en-us/cpp/code-quality/clang-tidy?view=msvc-170) and how it can handle the [MS specific C++ extensions](https://clang.llvm.org/docs/MSVCCompatibility.html) (SEH `__try`/`__except` keywords, etc.), but it seems that all this is now possible,

I think `clang-tidy` is also worth to try out...
